### PR TITLE
oneDNN FullyConnected weight caching & refactor

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
Caching mechanism for FC
BRGEMM always enabled for int8
Refactor Forward function using smaller functions

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, clang, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, clang, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Review by DominikaJedynak — APPROVED** [review]

File: src/operator/subgraph/dnnl/dnnl_fc_property.h [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
 #include <string>
 #include <vector>
+#include <random>
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply src/operator/subgraph/dnnl/dnnl_fc_property.h]: now not, done

File: src/operator/subgraph/dnnl/dnnl_fc.cc [hunk 2]
```diff
@@ -63,11 +65,26 @@ class SgDNNLFCOp {
   }
 
  private:
+  enum { kDataMin = 0, kDataMax, kWeightMin, kWeightMax, kBiasMin, kBiasMax, kSumMin, kSumMax };
+
+  NDArray PrepareOutputWithSum(const NDArray& sum_input, const NDArray& output);
+  bool CheckInitializationConditions(const std::vector<NDArray>& inputs,
+                                     const std::vector<float>& minmaxvec,
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply src/operator/subgraph/dnnl/dnnl_fc.cc]: done

File: src/operator/subgraph/dnnl/dnnl_fc.cc [hunk 3]
```diff
@@ -112,104 +128,54 @@ void SgDNNLFCOp::Forward(const OpContext& ctx,
...
+  minmaxvec[kDataMin]   = 0.0f;
+  minmaxvec[kDataMax]   = 0.0f;
+  minmaxvec[kWeightMin] = 0.0f;
+  minmaxvec[kWeightMax] = 0.0f;
+  minmaxvec[kBiasMin]   = 0.0f;
+  minmaxvec[kBiasMax]   = 0.0f;
 
-  const float sum_min   = idx.sum_min ? in_data[idx.sum_min].data().dptr<float>()[0] : 0.0;
-  const float sum_max   = idx.sum_max ? in_data[idx.sum_max].data().dptr<float>()[0] : 0.0;
+  minmaxvec[kSumMin]    = idx.sum_min ? in_data[idx.sum_min].data().dptr<float>()[0] : 0.0;
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply src/operator/subgraph/dnnl/dnnl_fc.cc]: done

**Comment by DominikaJedynak** [comment]
> @DominikaJedynak please check if I merged correctly your type fix

Yes, it seems to be ok.

**Review by bartekkuncer — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [all]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu, unix-gpu, centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Pull request merged**